### DISCUSSION TITLE
Layered ClassLoader issues

### DISCUSSION THREAD

**Discussion by lordofthejars** [langgraph4j/langgraph4j#27] [General]
Hi, I opened an issue in the Quarkus project about one problem I faced during the serialization process, and it seems that the problem is widely extended to any project that can use a layered ClassLoader (https://github.com/quarkusio/quarkus/issues/43111#event-14173788950)

I know that reworking this might be complicated here, but it is an issue that might be raised in the future.

**Comment by bsorrentino** [comment]
Hi @lordofthejars thanks for feedback

Do you know if also library (like Jackson) had the same problem ?

**Reply by lordofthejars** [reply]
The problem is in the serialization library. In Jackson, this should not be the problem.

What I was thinking is, why force that? For memory, I understand you need to clone the object to protect objects from external modification, but maybe you could use a copy-deep feature like https://commons.apache.org/proper/commons-beanutils/.

For memory saver, it is memory after all, so you can store the object itself. And then, for any other implementations, you can leave it to implementations to do it without forcing them to use any framework.

Maybe I am missing something, but maybe this would be the most agnostic way without classloader issues.

**Comment by bsorrentino** [comment]
Hi @lordofthejars  thanks for valuable feedback

Serialization seemed to me the best way to both clone and persist data ( also on file system ).

However I've implemented the [Serializer] outer the standard serialisation mechanism  and  the instantiation is explicit (not using the lookup in  classoladers). This approach requires bit of code but is safe and controlled. I'm adding the possibility to serialise in different format (eg. Json) using third-party library.

Deep-Copy could be an interesting alternative just in case


[Serializer]: https://bsorrentino.github.io/langgraph4j/apidocs/org/bsc/langgraph4j/serializer/Serializer.html

**Reply by lordofthejars** [reply]
Hi, any ideas about this? I face this problem, especially when running tests in Quarkus, because of classloader issues. I am not sure if this issue will impact other frameworks.

**Comment by bsorrentino** [comment]
Hi @lordofthejars

Are you using last version ? Could you share more infos ? (eg. Stacktrace )

Anyway I'm making serialisation strategy customisable ( ref. last SNAPSHOT )
